src/libraries/Native/Unix/System.Native/pal_networking.c

```diff
@@ -440,6 +429,135 @@ int32_t SystemNative_GetHostEntryForName(const uint8_t* address, HostEntry* entr
     return ret;
 }
 
+#if HAVE_GETADDRINFO_A
+struct GetAddrInfoAsyncState
+{
+    struct gaicb gai_request;
+    struct gaicb* gai_requests;
+    struct sigevent sigevent;
+
+    HostEntry* entry;
+    GetHostEntryForNameCallback callback;
+    char address[];
+};
+
+static void GetHostEntryForNameAsyncComplete(sigval_t context)
+{
+    struct GetAddrInfoAsyncState* state = (struct GetAddrInfoAsyncState*)context.sival_ptr;
+
+    atomic_thread_fence(memory_order_acquire);
+
+    GetHostEntryForNameCallback callback = state->callback;
+
+    int ret = ConvertGetAddrInfoAndGetNameInfoErrorsToPal(gai_error(&state->gai_request));
+
+    if (ret == 0)
+    {
+        const uint8_t* address = (const uint8_t*)state->address;
+        struct addrinfo* info = state->gai_request.ar_result;
+
+        ret = GetHostEntries(address, info, state->entry);
+    }
+
+    if (callback != NULL)
+    {
+        callback(state->entry, ret);
+    }
+
+    free(state);
+}
+#endif
+
+int32_t SystemNative_PlatformSupportsGetAddrInfoAsync()
+{
+    return HAVE_GETADDRINFO_A;
+}
+
+// Get all address families and the canonical name
+static const struct addrinfo s_hostEntryForNameHints = {.ai_family = AF_UNSPEC, .ai_flags = AI_CANONNAME};
+
+int32_t SystemNative_GetHostEntryForName(const uint8_t* address, HostEntry* entry)
+{
+    if (address == NULL || entry == NULL)
+    {
+        return GetAddrInfoErrorFlags_EAI_BADARG;
+    }
+
+    struct addrinfo* info = NULL;
+
+    int result = getaddrinfo((const char*)address, NULL, &s_hostEntryForNameHints, &info);
+    if (result != 0)
+    {
+        return ConvertGetAddrInfoAndGetNameInfoErrorsToPal(result);
+    }
+
+    return GetHostEntries(address, info, entry);
+}
+
+int32_t SystemNative_GetHostEntryForNameAsync(const uint8_t* address, HostEntry* entry, GetHostEntryForNameCallback callback)
+{
+#if HAVE_GETADDRINFO_A
+    if (address == NULL || entry == NULL)
+    {
+        return GetAddrInfoErrorFlags_EAI_BADARG;
```

Comment:
It seems like this is later consumed by GetSocketErrorForNativeError. Should the be errno e.g. EINVAL instead of  GetAddrInfoErrorFlags_EAI_BADARG?

Reply:
I think `GetAddrInfoErrorFlags_EAI_BADARG` is correct, as it is mapped to https://github.com/dotnet/runtime/blob/e29b83934e7ceef9905f82bd6328cbc72db2278b/src/libraries/Common/src/Interop/Unix/System.Native/Interop.HostEntry.cs#L14-L24